### ISSUE TITLE
Update failed – but it seems to have been successful

### ISSUE THREAD

**Issue by tastytea** [koreader/koreader#7581]
* KOReader version: **2021.04**, **2021.03**
* Device: **PB632** (PocketBook Touch HD 3)

#### Issue

After updating from 2021.03 to 2021.04, I hit the restart button in the dialog box. KOReader restarted, but reported that the update failed. It shows the new version in the help menu. I had the same issue when updating from 2021.02 to 2021.03 but didn't report it at the time.

#### Steps to reproduce

1. Update from 2021.03 to 2021.04
2. Restart KOReader

##### `crash.log` (if applicable)

[crash.log](https://github.com/koreader/koreader/files/6357174/crash.log)

These lines in the log are new after the update:

```
04/22/21-11:25:24 INFO  ./cache/fontlist/fontinfo.dat [string "./cache/fontlist/fontinfo.dat"]:1: unexpected symbol near '.' initializing it
04/22/21-11:25:26 INFO  Performing one-time migration for 20210306 (1/2)
04/22/21-11:25:26 INFO  Performing one-time migration for 20210306 (2/2)
04/22/21-11:25:26 INFO  Performing one-time migration for 20210330
ffi.load: sqlite3
04/22/21-11:25:26 INFO  Performing one-time migration for 20210404
04/22/21-11:25:26 INFO  Performing one-time migration for 20210409
04/22/21-11:25:26 INFO  Performing one-time migration for 20210412
04/22/21-11:25:26 WARN  os.rename: ./cache/calibre-libraries.lua: No such file or directory
04/22/21-11:25:26 WARN  os.rename: ./cache/calibre-books.dat: No such file or directory
04/22/21-11:25:26 INFO  Performing one-time migration for 20210414
04/22/21-11:25:26 WARN  os.remove: ./cache/calibre/books.dat: No such file or directory
```

**Comment by Frenzie** [comment]
> It shows the new version in the help menu.

Strictly speaking that does't mean the update succeeded, just that the update to the one specific file with the version string succeeded.

Is this the specific error message you're referring to btw? That is, after clicking "restart".

https://github.com/koreader/koreader/blob/ea3fa5c2c7ed7330e8e1e6ef838b780e7998ea51/platform/pocketbook/koreader.app#L35-L46

**Comment by LucyOne** [comment]
I have been getting the "Update failed" message on all my Pocketbooks since the update code started being displayed. The updates are still successful and everything works as it should.

**Comment by tastytea** [comment]
> Is this the specific error message you're referring to btw?
> […]
> Update failed :(
> KOReader may fail to function properly!

Yes.

**Comment by Frenzie** [comment]
Is the koreader.app thing updated properly too on PB or could that be an issue? That is, if it's older than ~January or February then it's outdated.

**Comment by tastytea** [comment]
koreader.app's mtime and ctime is 2021-04-22, same as koreader/reader.lua.

**Comment by Frenzie** [comment]
On a related note, on Kobo we do this:
https://github.com/koreader/koreader/blob/e4c9409f97099c36616e6f122657cb9b537d47bf/platform/kobo/koreader.sh#L104-L107
That seems to be missing from PB.

**Comment by tastytea** [comment]
These are the files that were installed/updated, according to `find applications/koreader -newermt "2021-04-22"`:

[new_files.txt](https://github.com/koreader/koreader/files/6357621/new_files.txt)

**Comment by NiLuJe** [comment]
@Frenzie: That's because of the whole "moving koreader.sh to a tmpfs for USBMS thing".

----

Someone will actually have to run an unpack manually in a shell to see what's what, but I suspect some kind of tar warnings affecting the return code. (IIRC, we're not root on PB, that might be a factor here).

**Comment by tastytea** [comment]
I copied http://ota.koreader.rocks/koreader-pocketbook-v2021.04.targz to the PocketBook, and then did this in a terminal (on the device):

```
cd /mnt/ext1
applications/koreader/tar" xf koreader*targz --no-same-permissions --no-same-owner
```

I got a lot of `Cannot change mode to rwxr-xr-x: Operation not permitted` errors. Log file: [tar.log](https://github.com/koreader/koreader/files/6360081/tar.log)

Inside PocketBook, we are user “reader”, the files in `/mnt/ext1/applications` are owned by “root”. It seems that everything that is copied via USB is owned by root.

**Comment by NiLuJe** [comment]
Ah, that would explain it.

I don't really see a way to automatically handle that sanely, short of doing this as root, which we can't ;).

e.g., if you do have root, `chown` the whole `koreader` folder to `reader` and this should magically go away.

But, otherwise, whether we keep `--no-same-permissions --no-same-owner` or not, the fact that stuff belongs to root and we're not probably means we're screwed.

(I mean, for science, you can try running this *without* `--no-same-permissions --no-same-owner`, but I suspect you'll see much of the same warnings).

**Comment by tastytea** [comment]
Ran it without `--no-same-permissions` and `--no-same-owner`, same result. Both arguments are the default for ordinary users, according to the manpage. I don't see any options to ignore the permissions completely. 🙁

**Comment by tastytea** [comment]
We *could* add something like `| grep -Ev '(Cannot change mode to rwxr-xr-x: Operation not permitted|Exiting with failure status)'` and only report failure if more than 0 lines remain.

**Comment by NiLuJe** [comment]
Does using `--no-overwrite-dir` help? The only failures appear to stem from existing *directories*, so, one would hope...

**Comment by tastytea** [comment]
I get the same output with `--no-overwrite-dir`.

**Comment by Frenzie** [comment]
Strange, the option seems to work for me when I try this example here: https://superuser.com/a/699231

**Comment by tastytea** [comment]
When I try the example with GNU tar 1.34, I get this output:

``` shellsession
[…]
$ LC_ALL=C tar -C b --no-overwrite-dir -xvf test.tar diffowner
diffowner/
tar: diffowner: Cannot change mode to rwxrwxrwx: Operation not permitted
diffowner/foo
tar: Exiting with failure status due to previous errors
$ echo $?
2
```

**Comment by tastytea** [comment]
I tried the example again, this time with GNU tar 1.32. It worked. I guess we found a bug in tar?

**Comment by NiLuJe** [comment]
It's a fix in tar 1.33 (https://git.savannah.gnu.org/cgit/tar.git/commit/?id=14d8fc718f0c872274b90991ee634b0cd8e1a6f0).

It will try to fix the permissions to ensures files can actually be created inside. Which makes sense. But we can't really do anything about stuff owned by root as a plain user ;o).

**Comment by NiLuJe** [comment]
Oh, well, stupid workaround it is, then.
